user-auth/app/src/ecall.rs

```diff
@@ -0,0 +1,128 @@
+use password_hash::rand_core::OsRng;
+use password_hash::{PasswordHashString, SaltString};
+use sgx_types::*;
+use thiserror::Error;
+
+// XXX: Certain assumptions are made relate to the hash context. See the doc
+// comment in the corresponding enclave code.
+const HASH_STRING_LENGTH: usize = 96;
+
+extern "C" {
+    fn hash_password(
+        eid: sgx_enclave_id_t,
+        retval: *mut sgx_status_t,
+        password: *const u8,
+        password_length: usize,
+        salt: *const u8,
+        user_id: *const u8,
+        user_id_length: usize,
+        hash_string: *mut u8,
+    ) -> sgx_status_t;
+    fn verify_password(
+        eid: sgx_enclave_id_t,
+        retval: *mut sgx_status_t,
+        password: *const u8,
+        password_length: usize,
+        user_id: *const u8,
+        user_id_length: usize,
+        stored_hash_string: *const u8,
+        verify_status: *mut VerifyPasswordStatus,
+    ) -> sgx_status_t;
+}
+
+#[derive(Debug, Error)]
+pub enum AuthError {
+    #[error("password does not match the given hash")]
+    EmptyUserId,
+    #[error("empty password supplied")]
+    EmptyPassword,
+    #[error("supplied hash is not a valid PHC hash string")]
+    InvalidHashString,
+}
+
+/// Compute a new password hash.
+pub fn safe_hash_password(
+    eid: sgx_enclave_id_t,
+    user_id: &[u8],
+    password: &str,
+) -> Result<String, AuthError> {
+    // generate a 16 byte salt using entropy supplied by the OS
+    let salt = SaltString::generate(&mut OsRng);
+    let mut hash_string_buf = [0u8; HASH_STRING_LENGTH];
+    let ecall_return = &mut sgx_status_t::default();
+    match (user_id, password) {
```

Comment:
This match feels artifical. I would instead...
```
if user_id.is_empty() {
    return Err(AuthError::EmptyUserId);
}
if password.is_empty() {
    return Err(AuthError::EmptyPassword);
}
...
```
At least that feels more readable.

A separate question is if this is the right place to do this validation. I would think that the frontend would, and we should simply send whatever we got forward?